_posts/2021-03-23-keccak-optimizer-bug.md

```diff
@@ -0,0 +1,119 @@
+---
+layout: post
+published: true
+title:  "Solidity Keccak Caching Bug"
+date:   2021-03-22
+author: Solidity Team
+category: Security Alerts
+---
+
+On March 20, 2021, a bug in Solidity's bytecode optimizer was found by differential fuzzing. The bug
+is fixed with [version 0.8.3](https://github.com/ethereum/solidity/releases/tag/v0.8.3) released on
+March 23, 2021. **The bug is present in all prior versions of Solidity.**
+
+We assigned the bug a severity level of "medium".
+
+## Technical Details
+
+**Summary**: The bytecode optimizer incorrectly re-used previously evaluated Keccak-256 hashes. You
+are unlikely to be affected if you do not compute Keccak-256 hashes in inline assembly.
+
+Solidity's bytecode optimizer has a step that can compute Keccak-256 hashes, if the contents of the
+memory, over which the Keccak-256 built-in function is invoked, are known during compilation time.
+This step also has a mechanism to determine that two Keccak-256 hashes are equal even if the values
+in memory are not known during compile time. This step had a bug where, the Keccak-256 hash of the
+same memory content, but of different sizes were considered equal. The following is a simple example
+that illustrates the bug:
+
+```solidity
+contract C {
+  function bug() public returns (uint a, uint b) {
+    assembly {
+      mstore(0, 0)
+      // The optimizer computes the value at compile time:
+      // 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
+      a := keccak256(0, 32)
+      // The optimizer incorrectly uses the cached value
+      // 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
+      // instead of 
+      // 0xe2b9f9f9430b05bfa9a3abd3bac9a181434d23a707ef1cde8bd25d30203538d8
+      b := keccak256(0, 23)
```

Comment:
```suggestion
      // The optimizer incorrectly uses the cached value
      // 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
      // and transforms the next line to
      // b := 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
      // instead of 
      // b := 0xe2b9f9f9430b05bfa9a3abd3bac9a181434d23a707ef1cde8bd25d30203538d8
      b := keccak256(0, 23)
```